src/Serializer/ItemNormalizer.php

```diff
@@ -76,10 +77,34 @@ private function updateObjectToPopulate(array $data, array &$context): void
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getResourceFromIri((string) $data['id'], $context + ['fetch_data' => true]);
         } catch (InvalidArgumentException) {
             $operation = $this->resourceMetadataCollectionFactory->create($context['resource_class'])->getOperation();
-            // todo: we could guess uri variables with the operation and the data instead of hardcoding id
-            $iri = $this->iriConverter->getIriFromResource($context['resource_class'], UrlGeneratorInterface::ABS_PATH, $operation, ['uri_variables' => ['id' => $data['id']]]);
+            $uriVariables = $this->guessUriVariables($data, $operation, $context);
+            $iri = $this->iriConverter->getIriFromResource($context['resource_class'], UrlGeneratorInterface::ABS_PATH, $operation, ['uri_variables' => $uriVariables]);
 
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getResourceFromIri($iri, ['fetch_data' => true]);
         }
     }
+
+    private function guessUriVariables(array $data, $operation, array $context): array
+    {
+        if (false === isset($context['uri_variables'])) {
```

Comment:
```suggestion
        if ($context['uri_variables'] ?? false) {
```

Reply:
@soyuka , I don't think it will work with this suggestion...

If `$context['uri_variables']` is set and has any value, then it will enter the `if`, doesn't it?

And we don't want this. We want it enter the `if` only if the key `uri_variables` doesn't exist (or is `null`, `false`, etc.).

This suggestion is, in practice, the opposite of what we want to achieve.

WDYT?

PS
Also the tests will fail with this modification.

Reply:
at least `!isset` then?

Reply:
@soyuka , ok, if you like it more 👍